grid/project-management/tasks/maps/atlassian-cloud.suma

```diff
@@ -0,0 +1,348 @@
+      }
+    }
+  }
+}
+
+operation ProcessIssues {
+  result = (() => {
+    // https://developer.atlassian.com/cloud/jira/platform/apis/document/structure/
+    const adfToMarkdown = (json) => {
+      let result = "";
+
+      const processNode = (node, parentType) => {
+        if (node.type === "text") {
+          let text = node.text;
+          if (node.marks) {
+            node.marks.forEach((mark) => {
+              if (mark.type === "strong") {
+                text = `**${text}**`;
+              } else if (mark.type === "em") {
+                text = `_${text}_`;
+              } else if (mark.type === "link") {
+                text = `[${text}](${mark.attrs.href})`;
+              } else if (mark.type === "underline") {
+                text = `__${text}__`;
+              } else if (mark.type === "strike") {
+                text = `~~${text}~~`;
+              } else if (mark.type === "code") {
+                text = `\`${text}\``;
+              }
+            });
+          }
+          return text;
+        } else if (node.type === "paragraph") {
+          return (
+            node.content.map((child) => processNode(child, node.type)).join("") +
+            "\n\n"
+          );
+        } else if (node.type === "orderedList") {
+          return node.content
+            .map((child) => processNode(child, node.type))
+            .join("");
+        } else if (node.type === "bulletList") {
+          return node.content
+            .map((child) => processNode(child, node.type))
+            .join("");
+        } else if (node.type === "listItem") {
+          const prefix = parentType === "orderedList" ? "1. " : "* ";
+          return (
+            prefix +
+            node.content.map((child) => processNode(child, node.type)).join("") +
+            "\n"
+          );
+        } else if (node.type === "mediaSingle") {
+          return `![media](${node.content[0].attrs.id})\n\n`;
+        } else if (node.type === "codeBlock") {
+          return (
+            "```\n" +
+            node.content.map((child) => processNode(child, node.type)).join("") +
+            "\n```\n\n"
+          );
+        } else if (node.type === "mention") {
+          return node.attrs.text;
+        } else if (node.type === "blockquote") {
+          return (
+            "> " +
+            node.content.map((child) => processNode(child, node.type)).join("") +
+            "\n"
+          );
+        } else if (node.type === "rule") {
+          return "---\n";
+        }
+
+        return "???\n";
+      }
+
+      json.content.forEach((node) => {
+        result += processNode(node);
+      });
+
+      return result;
+    }
+
+    return args.issues.map((issue) => {
+      return {
+        taskId: issue.id,
+        projectIds: [
+          issue.fields.project.id
+        ],
+        title: issue.fields.summary,
+        description: issue.fields.description ? adfToMarkdown(issue.fields.description) : undefined,
```

Comment:
question: Here you are mapping adf to markdown. What will happen if I perform ReadTask usecase and than patch description of returned task using UpdateTask usecase using the same markdown format?

Reply:
Good point - that'll just end up submitting the "markdown" description as plaintext.

I'm not sure how to proceed; converting ADF to pseudo-markdown seems reasonably simple, the other way around not so much.

Reply:
I suppose I could strip all formatting and only work with "paragraphs" on both sides; this loses some of the content, but is symmetrical at least.

Reply:
I do not like the inconsistency between the CreateTask, UpdateTask x ListTasks, ReadTask usecases.

I understand that support for markdown in CreateTask, UpdateTask maps is hard problem. It may be fine the way you have implemented it, as long as it is well documented in the profile. It depends what are the requirements of the client who requested the integration.

Reply:
>I do not like the inconsistency between the CreateTask, UpdateTask x ListTasks, ReadTask usecases.

Yeah, I do understand.

>It depends what are the requirements of the client who requested the integration.

Right, but we don't know those yet, and I'd rather have something out and iterate later.

In either case, the coming WASMy SDK may make all this much less awkward to deal with anyway.

---

Changed it so that we reduce everything to plain text, this works both ways.